src/window_identifier/wayland.rs

```diff
@@ -1,40 +1,57 @@
 use std::fmt;
 
 use wayland_client::{
-    protocol::{__interfaces::WL_SURFACE_INTERFACE, wl_surface::WlSurface},
+    protocol::{__interfaces::WL_SURFACE_INTERFACE, wl_registry, wl_surface::WlSurface},
     Proxy, QueueHandle,
 };
-use wayland_protocols::xdg::foreign::zv2::client::{
-    zxdg_exported_v2::{Event, ZxdgExportedV2},
-    zxdg_exporter_v2::ZxdgExporterV2,
+use wayland_protocols::xdg::foreign::{
+    zv1::client::{
+        zxdg_exported_v1::{self, ZxdgExportedV1},
+        zxdg_exporter_v1::ZxdgExporterV1,
+    },
+    zv2::client::{
+        zxdg_exported_v2::{self, ZxdgExportedV2},
+        zxdg_exporter_v2::ZxdgExporterV2,
+    },
 };
 
 use super::WindowIdentifierType;
 
+// Supported versions.
+const ZXDG_EXPORTER_V1: u32 = 1;
+const ZXDG_EXPORTER_V2: u32 = 1;
+
+#[derive(Debug)]
 pub struct WaylandWindowIdentifier {
-    exported: ZxdgExportedV2,
+    exported_v1: Option<ZxdgExportedV1>,
+    exported_v2: Option<ZxdgExportedV2>,
     type_: WindowIdentifierType,
 }
 
 impl WaylandWindowIdentifier {
-    pub fn new(surface: &WlSurface) -> Option<Self> {
-        match wayland_handle_export_from_surface(surface) {
-            Ok((exported, handle)) => Some(Self {
-                exported,
-                type_: WindowIdentifierType::Wayland(handle),
-            }),
-            _ => None,
-        }
+    pub async fn new(surface: &WlSurface) -> Option<Self> {
+        wayland_export_handle(surface).await
     }
 
-    pub fn from_raw(surface_ptr: *mut std::ffi::c_void) -> Option<Self> {
-        match wayland_handle_export(surface_ptr) {
-            Ok((exported, handle)) => Some(Self {
-                exported,
-                type_: WindowIdentifierType::Wayland(handle),
-            }),
-            _ => None,
-        }
+    pub async fn from_raw(
+        surface_ptr: *mut std::ffi::c_void,
+        display_ptr: *mut std::ffi::c_void,
+    ) -> Option<Self> {
+        let backend = unsafe {
+            wayland_backend::sys::client::Backend::from_foreign_display(display_ptr as *mut _)
+        };
+        let conn = wayland_client::Connection::from_backend(backend);
```

Comment:
I don't think we have much of an option to avoid this.

Plus the `from_foreign_display` will act as a guest backend, so all the handles will still belong to the same underlying display.